# Exemplos de Classificação em Ruby

### Notebook de Rubens Minoru Andako Bueno
#### [VAGAS.com](http://www.vagas.com.br/)

## Sumário

2. [Bibliotecas](#Bibliotecas)

3. [Desafio](#Desafio)

4. [Questões](#Questões)

5. [Passo 1: Visualizando os dados](#Passo-1:-Visualizando-os-dados)

6. [Passo 2: Pré-processamento](#Passo-2:-Pré-processamento)

7. [Passo 3: Classificação](#Passo-3:-Classificação)

## Bibliotecas

[[ Volte ao topo ]](#Sumário)

Os principais pacotes que nós iremos usar são:

* **daru**: Biblioteca para armazenamento, analise, manipulação e visualização de dados em Ruby.
* **nyaplot**: Biblioteca para gerar gráficos interativos em Ruby.
* **decisiontree**: Biblioteca que implementa o algoritmo ID3 (ganho de informação) para aprendizado de árvores de decisão.

## Desafio

[[ Volte ao topo ]](#Sumário)

Quais são os **atributos**?
- **sepal length**: Comprimento da sépala
- **sepal width**: Largura da sépala
- **petal length**: Comprimento da pétala
- **petal width**: Largura da pétala

![Descrição dos atributos](iris_with_labels.jpg)

Qual a **resposta**?
- **class**: Espécie da flor.

![Tipos de Íris](iris_types.png)

Vamos dar uma olhada nos dados:

In [2]:
require 'daru'

iris_data = Daru::DataFrame.from_csv('iris-data.csv')
iris_data.head(5)

"if(window['d3'] === undefined ||\n   window['Nyaplot'] === undefined){\n    var path = {\"d3\":\"http://d3js.org/d3.v3.min\",\"downloadable\":\"http://cdn.rawgit.com/domitry/d3-downloadable/master/d3-downloadable\"};\n\n\n\n    var shim = {\"d3\":{\"exports\":\"d3\"},\"downloadable\":{\"exports\":\"downloadable\"}};\n\n    require.config({paths: path, shim:shim});\n\n\nrequire(['d3'], function(d3){window['d3']=d3;console.log('finished loading d3');require(['downloadable'], function(downloadable){window['downloadable']=downloadable;console.log('finished loading downloadable');\n\n\tvar script = d3.select(\"head\")\n\t    .append(\"script\")\n\t    .attr(\"src\", \"http://cdn.rawgit.com/domitry/Nyaplotjs/master/release/nyaplot.js\")\n\t    .attr(\"async\", true);\n\n\tscript[0][0].onload = script[0][0].onreadystatechange = function(){\n\n\n\t    var event = document.createEvent(\"HTMLEvents\");\n\t    event.initEvent(\"load_nyaplot\",false,false);\n\t    window.dispatchEvent(event);\n\t

#<Daru::DataFrame(5x5)>
            sepal_leng sepal_widt petal_leng petal_widt      class
          0        5.1        3.5        1.4        0.2 Iris-setos
          1        4.9          3        1.4        0.2 Iris-setos
          2        4.7        3.2        1.3        0.2 Iris-setos
          3        4.6        3.1        1.5        0.2 Iris-setos
          4          5        3.6        1.4        0.2 Iris-setos

## Questões

[[ Volte ao topo ]](#Sumário)

Vamos fingir que você trabalha em uma empresa que deseja desenvolver um aplicativo voltado para amantes e que ajude a identificar o tipo de íris baseado somente em medidas fáceis que podem ser tiradas por um leigo.

Vamos primeiro buscar responder as seguintes perguntas:

* Existe relação entre as medidas e o tipo da flor?
* Existe algo errado dentro dos dados?
* É possível criar um modelo capaz de descobrir a espécie apenas com essas medidas?
* Com quanta certeza podemos prever?

## Passo 1: Visualizando os dados

[[ Volte ao topo ]](#Sumário)

In [3]:
iris_data.describe

#<Daru::DataFrame(5x4)>
            petal_leng petal_widt sepal_leng sepal_widt
      count        150        145        150        150
       mean 3.75866666 1.23655172 5.64462666 3.05466666
        std 1.76442041 0.75505829 1.31278140 0.43312299
        min          1        0.1      0.055          2
        max        6.9        2.5        7.9        4.4

Vamos visualizar esses dados de uma forma mais gráfica:

In [4]:
require 'nyaplot'

iris_data_valid = iris_data.clone_only_valid.to_nyaplotdf
graphic_colors = Nyaplot::Colors.qual

#<Nyaplot::Color:0x007f910f1ae0f8 @source=["rgb(251,180,174)", "rgb(179,205,227)", "rgb(204,235,197)", "rgb(222,203,228)", "rgb(254,217,166)", "rgb(255,255,204)", "rgb(229,216,189)", "rgb(253,218,236)", "rgb(242,242,242)"]>

In [5]:
plot = Nyaplot::Plot.new
sc = plot.add_with_df(iris_data_valid, :scatter, :petal_length_cm, :petal_width_cm)
plot.x_label("Petal Length")
plot.y_label("Petal Width")
sc.color(graphic_colors)
sc.fill_by(:class)
plot.show()

#<Nyaplot::Frame:0x007f910f19c510 @properties={:panes=>[#<Nyaplot::Plot:0x007f910f19cf60 @properties={:diagrams=>[#<Nyaplot::Diagram:0x007f910f19cd58 @properties={:type=>:scatter, :options=>{:x=>:petal_length_cm, :y=>:petal_width_cm, :color=>#<Nyaplot::Color:0x007f910f1ae0f8 @source=["rgb(251,180,174)", "rgb(179,205,227)", "rgb(204,235,197)", "rgb(222,203,228)", "rgb(254,217,166)", "rgb(255,255,204)", "rgb(229,216,189)", "rgb(253,218,236)", "rgb(242,242,242)"]>, :fill_by=>:class}, :data=>"8786904f-f82a-4c1f-bbf6-f6f84cde9708"}, @xrange=[1, 6.9], @yrange=[0.1, 2.5]>], :options=>{:x_label=>"Petal Length", :y_label=>"Petal Width", :zoom=>true, :width=>700, :xrange=>[1, 6.9], :yrange=>[0.1, 2.5]}}>], :data=>{"8786904f-f82a-4c1f-bbf6-f6f84cde9708"=>#<Nyaplot::DataFrame:0x007f910fd49ef8 @name="8786904f-f82a-4c1f-bbf6-f6f84cde9708", @rows=[{:sepal_length_cm=>5.1, :sepal_width_cm=>3.5, :petal_length_cm=>1.4, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.9, :sepal_width_cm=>3, :petal_length_cm=>1.4, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.7, :sepal_width_cm=>3.2, :petal_length_cm=>1.3, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.6, :sepal_width_cm=>3.1, :petal_length_cm=>1.5, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5, :sepal_width_cm=>3.6, :petal_length_cm=>1.4, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.4, :sepal_width_cm=>3.9, :petal_length_cm=>1.7, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.6, :sepal_width_cm=>3.4, :petal_length_cm=>1.4, :petal_width_cm=>0.3, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.8, :sepal_width_cm=>3, :petal_length_cm=>1.4, :petal_width_cm=>0.1, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.7, :sepal_width_cm=>3, :petal_length_cm=>1.1, :petal_width_cm=>0.1, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.8, :sepal_width_cm=>4, :petal_length_cm=>1.2, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.7, :sepal_width_cm=>4.4, :petal_length_cm=>1.5, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.4, :sepal_width_cm=>3.9, :petal_length_cm=>1.3, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.1, :sepal_width_cm=>3.5, :petal_length_cm=>1.4, :petal_width_cm=>0.3, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.7, :sepal_width_cm=>3.8, :petal_length_cm=>1.7, :petal_width_cm=>0.3, :class=>"Iris-setossa"}, {:sepal_length_cm=>5.1, :sepal_width_cm=>3.8, :petal_length_cm=>1.5, :petal_width_cm=>0.3, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.4, :sepal_width_cm=>3.4, :petal_length_cm=>1.7, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.1, :sepal_width_cm=>3.7, :petal_length_cm=>1.5, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.6, :sepal_width_cm=>3.6, :petal_length_cm=>1, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.1, :sepal_width_cm=>3.3, :petal_length_cm=>1.7, :petal_width_cm=>0.5, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.8, :sepal_width_cm=>3.4, :petal_length_cm=>1.9, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5, :sepal_width_cm=>3, :petal_length_cm=>1.6, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5, :sepal_width_cm=>3.4, :petal_length_cm=>1.6, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.2, :sepal_width_cm=>3.5, :petal_length_cm=>1.5, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.2, :sepal_width_cm=>3.4, :petal_length_cm=>1.4, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.7, :sepal_width_cm=>3.2, :petal_length_cm=>1.6, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.8, :sepal_width_cm=>3.1, :petal_length_cm=>1.6, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.4, :sepal_width_cm=>3.4, :petal_length_cm=>1.5, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.2, :sepal_width_cm=>4.1, :petal_length_cm=>1.5

In [6]:
iris_data_valid = iris_data.clone_only_valid.to_nyaplotdf

plot = Nyaplot::Plot.new
sc = plot.add_with_df(iris_data_valid, :scatter, :petal_length_cm, :sepal_length_cm)
plot.x_label("Petal Length")
plot.y_label("Sepal Length")
sc.color(graphic_colors)
sc.fill_by(:class)
plot.show()

#<Nyaplot::Frame:0x007f910fb6ff10 @properties={:panes=>[#<Nyaplot::Plot:0x007f910fb74858 @properties={:diagrams=>[#<Nyaplot::Diagram:0x007f910fb746a0 @properties={:type=>:scatter, :options=>{:x=>:petal_length_cm, :y=>:sepal_length_cm, :color=>#<Nyaplot::Color:0x007f910f1ae0f8 @source=["rgb(251,180,174)", "rgb(179,205,227)", "rgb(204,235,197)", "rgb(222,203,228)", "rgb(254,217,166)", "rgb(255,255,204)", "rgb(229,216,189)", "rgb(253,218,236)", "rgb(242,242,242)"]>, :fill_by=>:class}, :data=>"3703323d-e3eb-406a-8c72-df013d71ae94"}, @xrange=[1, 6.9], @yrange=[0.055, 7.9]>], :options=>{:x_label=>"Petal Length", :y_label=>"Sepal Length", :zoom=>true, :width=>700, :xrange=>[1, 6.9], :yrange=>[0.055, 7.9]}}>], :data=>{"3703323d-e3eb-406a-8c72-df013d71ae94"=>#<Nyaplot::DataFrame:0x007f910fb95800 @name="3703323d-e3eb-406a-8c72-df013d71ae94", @rows=[{:sepal_length_cm=>5.1, :sepal_width_cm=>3.5, :petal_length_cm=>1.4, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.9, :sepal_width_cm=>3, :petal_length_cm=>1.4, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.7, :sepal_width_cm=>3.2, :petal_length_cm=>1.3, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.6, :sepal_width_cm=>3.1, :petal_length_cm=>1.5, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5, :sepal_width_cm=>3.6, :petal_length_cm=>1.4, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.4, :sepal_width_cm=>3.9, :petal_length_cm=>1.7, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.6, :sepal_width_cm=>3.4, :petal_length_cm=>1.4, :petal_width_cm=>0.3, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.8, :sepal_width_cm=>3, :petal_length_cm=>1.4, :petal_width_cm=>0.1, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.7, :sepal_width_cm=>3, :petal_length_cm=>1.1, :petal_width_cm=>0.1, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.8, :sepal_width_cm=>4, :petal_length_cm=>1.2, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.7, :sepal_width_cm=>4.4, :petal_length_cm=>1.5, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.4, :sepal_width_cm=>3.9, :petal_length_cm=>1.3, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.1, :sepal_width_cm=>3.5, :petal_length_cm=>1.4, :petal_width_cm=>0.3, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.7, :sepal_width_cm=>3.8, :petal_length_cm=>1.7, :petal_width_cm=>0.3, :class=>"Iris-setossa"}, {:sepal_length_cm=>5.1, :sepal_width_cm=>3.8, :petal_length_cm=>1.5, :petal_width_cm=>0.3, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.4, :sepal_width_cm=>3.4, :petal_length_cm=>1.7, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.1, :sepal_width_cm=>3.7, :petal_length_cm=>1.5, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.6, :sepal_width_cm=>3.6, :petal_length_cm=>1, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.1, :sepal_width_cm=>3.3, :petal_length_cm=>1.7, :petal_width_cm=>0.5, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.8, :sepal_width_cm=>3.4, :petal_length_cm=>1.9, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5, :sepal_width_cm=>3, :petal_length_cm=>1.6, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5, :sepal_width_cm=>3.4, :petal_length_cm=>1.6, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.2, :sepal_width_cm=>3.5, :petal_length_cm=>1.5, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.2, :sepal_width_cm=>3.4, :petal_length_cm=>1.4, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.7, :sepal_width_cm=>3.2, :petal_length_cm=>1.6, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.8, :sepal_width_cm=>3.1, :petal_length_cm=>1.6, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.4, :sepal_width_cm=>3.4, :petal_length_cm=>1.5, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.2, :sepal_width_cm=>4.1, :petal_length_c

Baseado nas visualizações, podemos concluir:

1. Existem cinco classes enquanto deveria existir apenas três

2. Muitas entradas de `sepal_width_cm` para `Iris-setosa` são próximas de zero.

3. Muitas entradas de `sepal_length_cm` para `Iris-versicolor` são próximas de zero.

4. Alguns dados estão faltando.

## Passo 2: Pré-processamento

[[ Volte ao topo ]](#Sumário)

> Existem cinco classes enquanto deveria existir apenas três

Vamos arrumar o primeiro problema com os dados:

In [7]:
iris_data['class'].to_a.uniq()

["Iris-setosa", "Iris-setossa", "Iris-versicolor", "versicolor", "Iris-virginica"]

In [8]:
iris_data['class'] = iris_data.collect(:row) do |row|
  if row['class'] == 'versicolor'
    'Iris-versicolor'
  elsif row['class'] == 'Iris-setossa'
    'Iris-setosa'
  else
    row['class']
  end
end

#<Daru::Vector(150)>
           0 Iris-setosa
           1 Iris-setosa
           2 Iris-setosa
           3 Iris-setosa
           4 Iris-setosa
           5 Iris-setosa
           6 Iris-setosa
           7 Iris-setosa
           8 Iris-setosa
           9 Iris-setosa
          10 Iris-setosa
          11 Iris-setosa
          12 Iris-setosa
          13 Iris-setosa
          14 Iris-setosa
         ...         ...

> Muitas entradas de `sepal_width_cm` para `Iris-setosa` são próximas de zero.

No caso das entradas anômalas para `Iris-setosa`, as medidas foram erroneamente obtidas. O mais prático no momento é simplesmente remover esses dados:

In [9]:
iris_data = iris_data.where(iris_data['class'].not_eq('Iris-setosa').or(iris_data['sepal_width_cm'].gteq(2.5)))

#<Daru::DataFrame(149x5)>
            sepal_leng sepal_widt petal_leng petal_widt      class
          0        5.1        3.5        1.4        0.2 Iris-setos
          1        4.9          3        1.4        0.2 Iris-setos
          2        4.7        3.2        1.3        0.2 Iris-setos
          3        4.6        3.1        1.5        0.2 Iris-setos
          4          5        3.6        1.4        0.2 Iris-setos
          5        5.4        3.9        1.7        0.4 Iris-setos
          6        4.6        3.4        1.4        0.3 Iris-setos
          7          5        3.4        1.5        nil Iris-setos
          8        4.4        2.9        1.4        nil Iris-setos
          9        4.9        3.1        1.5        nil Iris-setos
         10        5.4        3.7        1.5        nil Iris-setos
         11        4.8        3.4        1.6        nil Iris-setos
         12        4.8          3        1.4        0.1 Iris-setos
         13        5.7          3        1.1        0.1 Iris-setos
         14        5.8          4        1.2        0.2 Iris-setos
        ...        ...        ...        ...        ...        ...

In [10]:
iris_data_valid = iris_data.clone_only_valid

hash_iris = iris_data_valid.where(iris_data['class'].eq('Iris-setosa'))['sepal_width_cm'].value_counts.to_h

plot = Nyaplot::Plot.new
sc = plot.add(:scatter, hash_iris.keys, hash_iris.values)
plot.x_label("Sepal Width")
plot.y_label("Count")
plot.show()

#<Nyaplot::Frame:0x007f91101aa380 @properties={:panes=>[#<Nyaplot::Plot:0x007f91101b0410 @properties={:diagrams=>[#<Nyaplot::Diagram:0x007f91101aaa88 @properties={:type=>:scatter, :options=>{:x=>"data0", :y=>"data1"}, :data=>"e3ed7584-ccef-482f-a916-04f3f17a2551"}, @xrange=[2.3, 4.4], @yrange=[1, 7]>], :options=>{:x_label=>"Sepal Width", :y_label=>"Count", :zoom=>true, :width=>700, :xrange=>[2.3, 4.4], :yrange=>[1, 7]}}>], :data=>{"e3ed7584-ccef-482f-a916-04f3f17a2551"=>#<Nyaplot::DataFrame:0x007f91101abe38 @name="e3ed7584-ccef-482f-a916-04f3f17a2551", @rows=[{:data0=>3.5, :data1=>6}, {:data0=>3, :data1=>6}, {:data0=>3.2, :data1=>7}, {:data0=>3.1, :data1=>5}, {:data0=>3.6, :data1=>2}, {:data0=>3.9, :data1=>2}, {:data0=>3.4, :data1=>7}, {:data0=>4, :data1=>1}, {:data0=>4.4, :data1=>1}, {:data0=>3.8, :data1=>4}, {:data0=>3.7, :data1=>2}, {:data0=>3.3, :data1=>2}, {:data0=>4.1, :data1=>1}, {:data0=>4.2, :data1=>1}, {:data0=>2.3, :data1=>1}, {:data0=>2.8, :data1=>1}]>}, :extension=>[]}>

> Muitas entradas de `sepal_length_cm` para `Iris-versicolor` são próximas de zero.

Alguns dos valores de `sepal_length_cm` da `Iris-versicolor` estão muito próximos de zero, como se eles estivessem sendo medidos em metros, invés de centímetros, então vamos corrigir esses dados.

In [11]:
iris_data.where(iris_data['class'].eq('Iris-versicolor').and(iris_data['sepal_length_cm'].lt(1.0)))

#<Daru::DataFrame(5x5)>
            sepal_leng sepal_widt petal_leng petal_widt      class
         77      0.067          3          5        1.7 Iris-versi
         78       0.06        2.9        4.5        1.5 Iris-versi
         79      0.057        2.6        3.5          1 Iris-versi
         80      0.055        2.4        3.8        1.1 Iris-versi
         81      0.055        2.4        3.7          1 Iris-versi

In [12]:
iris_data['sepal_length_cm'] = iris_data.collect(:row) do |row|
  if row['class'] == 'Iris-versicolor' && row['sepal_length_cm'] < 1.0
    row['sepal_length_cm'] * 100.0
  else
    row['sepal_length_cm']
  end
end

#<Daru::Vector(149)>
   0 5.1
   1 4.9
   2 4.7
   3 4.6
   4   5
   5 5.4
   6 4.6
   7   5
   8 4.4
   9 4.9
  10 5.4
  11 4.8
  12 4.8
  13 5.7
  14 5.8
 ... ...

> Alguns dados estão faltando.

Muitos dos `petal_width_cm` da `Iris-setosa` está entre 0.2 e 0.3, então no momento o mais prático será preencher os valores faltantes com a média das outras observações:

In [13]:
average_petal_width_cm = iris_data.where(iris_data['class'].eq('Iris-setosa'))['petal_width_cm'].mean()

iris_data['petal_width_cm'] = iris_data.collect(:row) do |row|
  if row['class'] == 'Iris-setosa' and row['petal_width_cm'] == nil
    average_petal_width_cm
  else
    row['petal_width_cm']
  end
end

#<Daru::Vector(149)>
                   0                 0.2
                   1                 0.2
                   2                 0.2
                   3                 0.2
                   4                 0.2
                   5                 0.4
                   6                 0.3
                   7 0.24999999999999997
                   8 0.24999999999999997
                   9 0.24999999999999997
                  10 0.24999999999999997
                  11 0.24999999999999997
                  12                 0.1
                  13                 0.1
                  14                 0.2
                 ...                 ...

Vamos visualizar como ficam os gráficos agora:

In [14]:
iris_data_valid = iris_data.clone_only_valid.to_nyaplotdf

plot = Nyaplot::Plot.new
sc = plot.add_with_df(iris_data_valid, :scatter, :petal_length_cm, :sepal_length_cm)
plot.x_label("Petal Length")
plot.y_label("Sepal Length")
sc.color(graphic_colors)
sc.fill_by(:class)
plot.show()

#<Nyaplot::Frame:0x007f910e9b2f20 @properties={:panes=>[#<Nyaplot::Plot:0x007f910e9b3ad8 @properties={:diagrams=>[#<Nyaplot::Diagram:0x007f910e9b3880 @properties={:type=>:scatter, :options=>{:x=>:petal_length_cm, :y=>:sepal_length_cm, :color=>#<Nyaplot::Color:0x007f910f1ae0f8 @source=["rgb(251,180,174)", "rgb(179,205,227)", "rgb(204,235,197)", "rgb(222,203,228)", "rgb(254,217,166)", "rgb(255,255,204)", "rgb(229,216,189)", "rgb(253,218,236)", "rgb(242,242,242)"]>, :fill_by=>:class}, :data=>"e1e528e5-8fab-4b06-a3a8-0259411365ea"}, @xrange=[1, 6.9], @yrange=[4.4, 7.9]>], :options=>{:x_label=>"Petal Length", :y_label=>"Sepal Length", :zoom=>true, :width=>700, :xrange=>[1, 6.9], :yrange=>[4.4, 7.9]}}>], :data=>{"e1e528e5-8fab-4b06-a3a8-0259411365ea"=>#<Nyaplot::DataFrame:0x007f910e9e8170 @name="e1e528e5-8fab-4b06-a3a8-0259411365ea", @rows=[{:sepal_length_cm=>5.1, :sepal_width_cm=>3.5, :petal_length_cm=>1.4, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.9, :sepal_width_cm=>3, :petal_length_cm=>1.4, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.7, :sepal_width_cm=>3.2, :petal_length_cm=>1.3, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.6, :sepal_width_cm=>3.1, :petal_length_cm=>1.5, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5, :sepal_width_cm=>3.6, :petal_length_cm=>1.4, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.4, :sepal_width_cm=>3.9, :petal_length_cm=>1.7, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.6, :sepal_width_cm=>3.4, :petal_length_cm=>1.4, :petal_width_cm=>0.3, :class=>"Iris-setosa"}, {:sepal_length_cm=>5, :sepal_width_cm=>3.4, :petal_length_cm=>1.5, :petal_width_cm=>0.24999999999999997, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.4, :sepal_width_cm=>2.9, :petal_length_cm=>1.4, :petal_width_cm=>0.24999999999999997, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.9, :sepal_width_cm=>3.1, :petal_length_cm=>1.5, :petal_width_cm=>0.24999999999999997, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.4, :sepal_width_cm=>3.7, :petal_length_cm=>1.5, :petal_width_cm=>0.24999999999999997, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.8, :sepal_width_cm=>3.4, :petal_length_cm=>1.6, :petal_width_cm=>0.24999999999999997, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.8, :sepal_width_cm=>3, :petal_length_cm=>1.4, :petal_width_cm=>0.1, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.7, :sepal_width_cm=>3, :petal_length_cm=>1.1, :petal_width_cm=>0.1, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.8, :sepal_width_cm=>4, :petal_length_cm=>1.2, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.7, :sepal_width_cm=>4.4, :petal_length_cm=>1.5, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.4, :sepal_width_cm=>3.9, :petal_length_cm=>1.3, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.1, :sepal_width_cm=>3.5, :petal_length_cm=>1.4, :petal_width_cm=>0.3, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.7, :sepal_width_cm=>3.8, :petal_length_cm=>1.7, :petal_width_cm=>0.3, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.1, :sepal_width_cm=>3.8, :petal_length_cm=>1.5, :petal_width_cm=>0.3, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.4, :sepal_width_cm=>3.4, :petal_length_cm=>1.7, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.1, :sepal_width_cm=>3.7, :petal_length_cm=>1.5, :petal_width_cm=>0.4, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.6, :sepal_width_cm=>3.6, :petal_length_cm=>1, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5.1, :sepal_width_cm=>3.3, :petal_length_cm=>1.7, :petal_width_cm=>0.5, :class=>"Iris-setosa"}, {:sepal_length_cm=>4.8, :sepal_width_cm=>3.4, :petal_length_cm=>1.9, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5, :sepal_width_cm=>3, :petal_length_cm=>1.6, :petal_width_cm=>0.2, :class=>"Iris-setosa"}, {:sepal_length_cm=>5, :sepal_width_cm=>3.4, :petal_length_cm=>1.6, :petal_width_cm=>0.4, :class=>"Iri

## Passo 3: Classificação

[[ Volte ao topo ]](#Sumário)

Todo esse trabalho e nós ainda não modelamos os dados. Mas lembre-se: **Dados ruins levam a modelos ruins.**. Sempre visualize seus dados.

<hr />
Classificadores baseados em Árvores de Decisão são simples em teoria. Na forma mais simples, esses classificadores perguntam uma série de questões Sim/Não sobre o dados, até que ele consiga classificar os dados perfeitamente ou simplesmente não consiga mais diferenciar os dados.

Mas antes de treinar o classificador, devemos separar os dados em:

- **Conjunto de treinamento** subconjunto aleatório de dados usados para o treinamento

- **Conjunto de teste** subconjunto aleatório de dados usados para avaliar o modelo

Vamos separar nossos dados então:

In [17]:
require 'decisiontree'

def split_validation(dataset, split = 0.8)
  training_size = (dataset.length * split).floor
  
  [dataset[(0..training_size)], dataset[(training_size+1..dataset.length)]]
end

dataset = iris_data.map(:row) { |a| a.to_a }.shuffle
training_set, test_set = split_validation(dataset) 

[[[5.2, 3.5, 1.5, 0.2, "Iris-setosa"], [6.9, 3.2, 5.7, 2.3, "Iris-virginica"], [6.5, 3.2, 5.1, 2, "Iris-virginica"], [5.1, 3.8, 1.9, 0.4, "Iris-setosa"], [5.7, 2.9, 4.2, 1.3, "Iris-versicolor"], [5.6, 2.5, 3.9, 1.1, "Iris-versicolor"], [5.6, 3, 4.1, 1.3, "Iris-versicolor"], [6.4, 3.2, 4.5, 1.5, "Iris-versicolor"], [5.5, 2.4, 3.7, 1, "Iris-versicolor"], [7.3, 2.9, 6.3, 1.8, "Iris-virginica"], [6.1, 2.8, 4, 1.3, "Iris-versicolor"], [6.4, 2.8, 5.6, 2.1, "Iris-virginica"], [5.8, 2.7, 5.1, 1.9, "Iris-virginica"], [6.2, 3.4, 5.4, 2.3, "Iris-virginica"], [6.7, 3.1, 4.7, 1.5, "Iris-versicolor"], [6.8, 3, 5.5, 2.1, "Iris-virginica"], [5.5, 2.4, 3.8, 1.1, "Iris-versicolor"], [5.6, 3, 4.5, 1.5, "Iris-versicolor"], [6, 3, 4.8, 1.8, "Iris-virginica"], [5.3, 3.7, 1.5, 0.2, "Iris-setosa"], [6.8, 2.8, 4.8, 1.4, "Iris-versicolor"], [6.1, 2.9, 4.7, 1.4, "Iris-versicolor"], [7.2, 3.6, 6.1, 2.5, "Iris-virginica"], [6.0, 2.9, 4.5, 1.5, "Iris-versicolor"], [6.1, 3, 4.9, 1.8, "Iris-virginica"], [5.4, 3, 4.5,

Agora, treinamos o classificador com o conjunto de treinamento e vemos no conjunto de testes, quantos acertos ele obteve:

In [18]:
def precision(dataset, model)
  truth_table = dataset.map do |test|
    value = test[-1]
    predict = model.predict(test)

    puts("#{predict} (Actual: #{value}))")
    value == predict
  end.group_by { |x| x }.each_with_object({}) { |(k, v), acc| acc[k] = v.count }
  
  truth_table[true].to_f / (truth_table[false].to_f + truth_table[true].to_f)
end

attributes = {
  sepal_length_cm: :continuous,
  sepal_width_cm: :continuous,
  petal_length_cm: :continuous,
  petal_width_cm: :continuous
}

dec_tree = DecisionTree::ID3Tree.new(attributes.keys.map(&:to_s), training_set, 'Iris-setosa', attributes)
dec_tree.train

precision(test_set, dec_tree)

Iris-virginica (Actual: Iris-virginica))
Iris-setosa (Actual: Iris-setosa))
Iris-setosa (Actual: Iris-setosa))
Iris-setosa (Actual: Iris-setosa))
Iris-virginica (Actual: Iris-virginica))
Iris-virginica (Actual: Iris-virginica))
Iris-virginica (Actual: Iris-virginica))
Iris-setosa (Actual: Iris-setosa))
Iris-virginica (Actual: Iris-virginica))
Iris-versicolor (Actual: Iris-versicolor))
Iris-virginica (Actual: Iris-virginica))
Iris-virginica (Actual: Iris-virginica))
Iris-setosa (Actual: Iris-setosa))
Iris-versicolor (Actual: Iris-versicolor))
Iris-versicolor (Actual: Iris-versicolor))
Iris-versicolor (Actual: Iris-virginica))
Iris-setosa (Actual: Iris-setosa))
Iris-versicolor (Actual: Iris-versicolor))
Iris-virginica (Actual: Iris-virginica))
Iris-setosa (Actual: Iris-setosa))
Iris-virginica (Actual: Iris-virginica))
Iris-versicolor (Actual: Iris-versicolor))
Iris-virginica (Actual: Iris-virginica))
Iris-versicolor (Actual: Iris-versicolor))
Iris-virginica (Actual: Iris-virginica))
Iris

0.9655172413793104

**Muito bem!** Nosso modelo consegue atingir até 97% de precisão sem muito esforço. Dado que os conjuntos são aleatórios, nosso modelo pode atingir algo entre 80% e 100% de precisão.